Std/Data/MLList/Meta.lean

```diff
@@ -0,0 +1,238 @@
+  .squash fun _ => (·.2) <$> runGreedily jobs
+
+
+end Lean.Meta.MetaM
+
+namespace Lean.Elab.Term.TermElabM
+
+/--
+Given a monadic value in `TermElabM`, creates a task that runs it in the current state,
+returning
+* a cancellation hook and
+* a monadic value with the cached result (and subsequent state as it was after running).
+-/
+def asTask (t : TermElabM α) : TermElabM (BaseIO Unit × Task (TermElabM α)) := do
+  let (cancel, task) ← (t.run (← read) (← get)).asTask
+  return (cancel, task.map (prio := .max)
+    fun c : MetaM (α × Term.State) => do let (a, s) ← c; set s; pure a)
+
+/--
+Given a monadic value in `TermElabM`, creates a task that runs it in the current state,
+returning a monadic value with the cached result (and subsequent state as it was after running).
+-/
+def asTask' (t : TermElabM α) : TermElabM (Task (TermElabM α)) := (·.2) <$> asTask t
+
+/--
+Given a list of monadic values in `TermElabM`, runs them all as tasks,
+and returns
+* a cancellation hook and
+* the monadic lazy list which returns the values in the order they complete.
+
+See the doc-string for `IO.runGreedily` for details about the cancellation hook behaviour.
+-/
+def runGreedily (jobs : List (TermElabM α)) : TermElabM (BaseIO Unit × MLList TermElabM α) := do
+  let (cancels, tasks) := (← jobs.mapM asTask).unzip
+  return (cancels.forM id, .squash fun _ => return MLList.ofTaskList tasks |>.liftM.mapM id)
+
+/--
+Variant of `TermElabM.runGreedily` without a cancellation hook.
+-/
+def runGreedily' (jobs : List (TermElabM α)) : MLList TermElabM α :=
+  .squash fun _ => (·.2) <$> runGreedily jobs
+
+end Lean.Elab.Term.TermElabM
+
+namespace Lean.Elab.Tactic.TacticM
+
+/--
+Given a monadic value in `TacticM`, creates a task that runs it in the current state,
+returning
+* a cancellation hook and
+* a monadic value with the cached result (and subsequent state as it was after running).
+-/
+def asTask (t : TacticM α) : TacticM (BaseIO Unit × Task (TacticM α)) := do
+  let (cancel, task) ← (t (← read) |>.run (← get)).asTask
+  return (cancel, task.map (prio := .max)
+    fun c : TermElabM (α × Tactic.State) => do let (a, s) ← c; set s; pure a)
+
+/--
+Given a monadic value in `TacticM`, creates a task that runs it in the current state,
+returning a monadic value with the cached result (and subsequent state as it was after running).
+-/
+def asTask' (t : TacticM α) : TacticM (Task (TacticM α)) := (·.2) <$> asTask t
+
+/--
+Given a list of monadic values in `TacticM`, runs them all as tasks,
+and returns
+* a cancellation hook and
+* the monadic lazy list which returns the values in the order they complete.
+
+See the doc-string for `IO.runGreedily` for details about the cancellation hook behaviour.
+-/
+def runGreedily (jobs : List (TacticM α)) : TacticM (BaseIO Unit × MLList TacticM α) := do
+  let (cancels, tasks) := (← jobs.mapM asTask).unzip
+  return (cancels.forM id, .squash fun _ => return MLList.ofTaskList tasks |>.liftM.mapM id)
```

Comment:
What is. the intended semantics of `runGreedily`?  When you get a value from the tactic list, does retrieving it affect the tactic state?

Perhaps the `MLList` should have type `MLList BaseIO (Except Exception (α × Tactic.SavedState)`?

Reply:
Oh I wanted to see if I could avoid the need to double the number of tasks (via Task.map) and did that in [this branch](https://github.com/leanprover/std4/tree/jhx/MLList.Meta).

That preserves the behavior of restoring state implicitly though which I am not sure if the right approach.

Reply:
The intention is that the state is exposed via the MLList, so one could obtain `α × Tactic.SavedState` just by `.mapM fun a => return (a, ← saveState)`.

I can add a doc-string explaining this and a test.

Reply:
I added a test, and then tried merging your branch. Your branch failed the test, with a curiously strange error message: with a goal of `Nat`, `exact 0` failed with `failed to synthesize instance OfNat (_ : True) 0`! Something messed up in the metavar context, I think. So I haven't commited the merge of your branch.

Reply:
How important is avoiding the extra tasks? I can try again at this, or we could not worry. :-)

Reply:
I think this interface is not intuitive to me due to backtracking.  It's counter-intuitive to me, for example, that I could create a meta variable, get a value from a MLList generated by `runGreedily` in Meta and then the variable is no longer valid (because of `restoreState`).

Perhaps this patch should introduce a notion of a suspended computation (perhaps as a class) that can be resumed?  The suspended computation would have the state+context.  It's not quite as concise, but would make the code more clear.